<a href="https://colab.research.google.com/github/Nolberto-Rivera/MAT306-EntregablesNRivera/blob/main/docs/labs/lab_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/fralfaro/MAT306/blob/main/docs/labs/lab_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# MAT306 - Laboratorio N°03





**Objetivo**: Aplicar técnicas avanzadas de manipulación y análisis de datos con pandas sobre un conjunto real de datos de contenido de Netflix, reforzando buenas prácticas y métodos eficientes sin recurrir a `groupby`, `merge`, `pivot`, ni `join`.



**Dataset**:

Trabajaremos con el archivo `netflix_titles.csv`, que contiene información sobre los títulos disponibles en la plataforma Netflix hasta el año 2021.

| Variable       | Clase     | Descripción                                                                 |
|----------------|-----------|------------------------------------------------------------------------------|
| show_id        | caracter  | Identificador único del título en el catálogo de Netflix.                   |
| type           | caracter  | Tipo de contenido: 'Movie' o 'TV Show'.                                     |
| title          | caracter  | Título del contenido.                                                       |
| director       | caracter  | Nombre del director (puede ser nulo).                                       |
| cast           | caracter  | Lista de actores principales (puede ser nulo).                              |
| country        | caracter  | País o países donde se produjo el contenido.                                |
| date_added     | fecha     | Fecha en la que el título fue agregado al catálogo de Netflix.              |
| release_year   | entero    | Año de lanzamiento original del título.                                     |
| rating         | caracter  | Clasificación por edad (por ejemplo: 'PG-13', 'TV-MA').                      |
| duration       | caracter  | Duración del contenido (minutos o número de temporadas para series).        |
| listed_in      | caracter  | Categorías o géneros en los que está clasificado el contenido.              |
| description    | caracter  | Breve sinopsis del contenido.                                               |




In [1]:
import pandas as pd

# Cargar datos
df = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT306/main/docs/labs/data/netflix_titles.csv')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...


### Parte 1: Limpieza y preparación

1. Revisar y describir el dataset:

   * ¿Cuántas filas y columnas tiene?
   * ¿Qué tipos de datos hay?
   * ¿Cuántos valores nulos hay por columna?

2. Transformar la columna `date_added` a tipo fecha.

3. Crear columnas auxiliares con `assign`:

   * Año (`year_added`)
   * Mes (`month_added`)



**Respuesta PARTE 1:**

1. Revisar y describir el dataset:

   * ¿Cuántas filas y columnas tiene?
   * ¿Qué tipos de datos hay?
   * ¿Cuántos valores nulos hay por columna?

Podemos obtener todas las respuestas a las preguntas del ítem (1.) mediante el método `.info()`.

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


Vemos inmediatamente que:
    
* El dataset tiene 8807 columnas por 12 columnas.
* La columna `release_year` tiene tipo `int64`, mientras que el resto tiene tipo `object`.
* Los elementos nulos serán los elementos que faltan a los no-nulos para completar 8807 en cada columna. Así:
    * La columna `director` tiene 8807 - 6173 = 2634 elementos nulos.
    * La columna `cast` tiene 8807 - 7982 = 825 elementos nulos.
    * La columna `country` tiene 8807 - 7976 = 831 elementos nulos.
    * La columna `date_added` tiene 8807 - 8797 = 10 elementos nulos.
    * La columna `rating` tiene 8807 - 8803 = 4 elementos nulos.
    * La columna `duration` tiene 8807 - 8804 = 3 elementos nulos.

2. Transformar la columna `date_added` a tipo fecha.

3. Crear columnas auxiliares con `assign`:

   * Año (`year_added`)
   * Mes (`month_added`)

Notemos que `date_added` tiene 10 elementos nulos.

In [7]:
df.loc[df['date_added'].isnull()]

show_id     type                                        title director  \
6066   s6067  TV Show  A Young Doctor's Notebook and Other Stories      NaN   
6174   s6175  TV Show              Anthony Bourdain: Parts Unknown      NaN   
6795   s6796  TV Show                                      Frasier      NaN   
6806   s6807  TV Show                                      Friends      NaN   
6901   s6902  TV Show                              Gunslinger Girl      NaN   
7196   s7197  TV Show                                     Kikoriki      NaN   
7254   s7255  TV Show                          La Familia P. Luche      NaN   
7406   s7407  TV Show                                        Maron      NaN   
7847   s7848  TV Show                                 Red vs. Blue      NaN   
8182   s8183  TV Show                 The Adventures of Figaro Pho      NaN   

                                                   cast         country  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...  United Kingdom   
6174                                   Anthony Bourdain   United States   
6795  Kelsey Grammer, Jane Leeves, David Hyde Pierce...   United States   
6806  Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   United States   
6901  Yuuka Nanri, Kanako Mitsuhashi, Eri Sendai, Am...           Japan   
7196                                      Igor Dmitriev             NaN   
7254  Eugenio Derbez, Consuelo Duval, Luis Manuel Áv...   United States   
7406  Marc Maron, Judd Hirsch, Josh Brener, Nora Zeh...   United States   
7847  Burnie Burns, Jason Saldaña, Gustavo Sorola, G...   United States   
8182  Luke Jurevicius, Craig Behenna, Charlotte Haml...       Australia   

     date_added  release_year rating    duration  \
6066        NaN          2013  TV-MA   2 Seasons   
6174        NaN          2018  TV-PG   5 Seasons   
6795        NaN          2003  TV-PG  11 Seasons   
6806        NaN          2003  TV-14  10 Seasons   
6901        NaN          2008  TV-14   2 Seasons   
7196        NaN          2010   TV-Y   2 Seasons   
7254        NaN          2012  TV-14   3 Seasons   
7406        NaN          2016  TV-MA   4 Seasons   
7847        NaN          2015     NR  13 Seasons   
8182        NaN          2015  TV-Y7   2 Seasons   

                                              listed_in  \
6066           British TV Shows, TV Comedies, TV Dramas   
6174                                         Docuseries   
6795                     Classic & Cult TV, TV Comedies   
6806                     Classic & Cult TV, TV Comedies   
6901                       Anime Series, Crime TV Shows   
7196                                           Kids' TV   
7254  International TV Shows, Spanish-Language TV Sh...   
7406                                        TV Comedies   
7847  TV Action & Adventure, TV Comedies, TV Sci-Fi ...   
8182                              Kids' TV, TV Comedies   

                                            description  
6066  Set during the Russian Revolution, this comic ...  
6174  This CNN original series has chef Anthony Bour...  
6795  Frasier Crane is a snooty but lovable Seattle ...  
6806  This hit sitcom follows the merry misadventure...  
6901  On the surface, the Social Welfare Agency appe...  
7196  A wacky rabbit and his gang of animal pals hav...  
7254  This irreverent sitcom featues Ludovico, Feder...  
7406  Marc Maron stars as Marc Maron, who interviews...  
7847  This parody of first-person shooter games, mil...  
8182  Imagine your worst fears, then multiply them: ...

Al simplemente aplicar `pd.to_datetime` (con la opción `coerce`, para encargarse de cualquier error de formato), obtnemos lo siguiente.

In [9]:
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')
df.loc[df['date_added'].isnull()]

show_id     type                                        title  \
6066   s6067  TV Show  A Young Doctor's Notebook and Other Stories   
6079   s6080  TV Show                              Abnormal Summit   
6174   s6175  TV Show              Anthony Bourdain: Parts Unknown   
6177   s6178  TV Show                                     忍者ハットリくん   
6213   s6214  TV Show                                Bad Education   
...      ...      ...                                          ...   
8539   s8540  TV Show                                   The Tudors   
8557   s8558  TV Show                                The West Wing   
8684   s8685  TV Show                                      Vroomiz   
8712   s8713  TV Show                   Weird Wonders of the World   
8755   s8756  TV Show                            Women Behind Bars   

                     director  \
6066                      NaN   
6079  Jung-ah Im, Seung-uk Jo   
6174                      NaN   
6177                      NaN   
6213                      NaN   
...                       ...   
8539                      NaN   
8557                      NaN   
8684                      NaN   
8712                      NaN   
8755                      NaN   

                                                   cast  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...   
6079           Hyun-moo Jun, Si-kyung Sung, Se-yoon Yoo   
6174                                   Anthony Bourdain   
6177                                                NaN   
6213  Jack Whitehall, Mathew Horne, Sarah Solemani, ...   
...                                                 ...   
8539  Jonathan Rhys Meyers, Henry Cavill, James Frai...   
8557  Martin Sheen, Rob Lowe, Allison Janney, John S...   
8684  Joon-seok Song, Jeong-hwa Yang, Sang-hyun Um, ...   
8712                                      Chris Packham   
8755                                                NaN   

                                             country date_added  release_year  \
6066                                  United Kingdom        NaT          2013   
6079                                     South Korea        NaT          2017   
6174                                   United States        NaT          2018   
6177                                           Japan        NaT          2012   
6213                                  United Kingdom        NaT          2014   
...                                              ...        ...           ...   
8539  Ireland, Canada, United States, United Kingdom        NaT          2010   
8557                                   United States        NaT          2005   
8684                                     South Korea        NaT          2016   
8712                                  United Kingdom        NaT          2016   
8755                                   United States        NaT          2010   

     rating   duration                                          listed_in  \
6066  TV-MA  2 Seasons           British TV Shows, TV Comedies, TV Dramas   
6079  TV-PG  2 Seasons  International TV Shows, Korean TV Shows, Stand...   
6174  TV-PG  5 Seasons                                         Docuseries   
6177  TV-Y7  2 Seasons                             Anime Series, Kids' TV   
6213  TV-MA  3 Seasons                      British TV Shows, TV Comedies   
...     ...        ...                                                ...   
8539  TV-MA  4 Seasons                                          TV Dramas   
8557  TV-14  7 Seasons                                          TV Dramas   
8684   TV-Y  3 Seasons                          Kids' TV, Korean TV Shows   
8712  TV-PG  2 Seasons  British TV Shows, Docuseries, Science & Nature TV   
8755  TV-14  3 Seasons                         Crime TV Shows, Docuseries   

                                            description  
6066  Set during the Russian Revolution, this comic ...  
6079  Led by a trio of Korean celebs, a mult

Así también, al crear las columnas solicitadas, ocupando `assign`, se obtiene lo siguiente.

In [10]:
df = df.assign(
     year_added = df['date_added'].dt.year,
    month_added = df['date_added'].dt.month
)
df.loc[df['date_added'].isnull()]

show_id     type                                        title  \
6066   s6067  TV Show  A Young Doctor's Notebook and Other Stories   
6079   s6080  TV Show                              Abnormal Summit   
6174   s6175  TV Show              Anthony Bourdain: Parts Unknown   
6177   s6178  TV Show                                     忍者ハットリくん   
6213   s6214  TV Show                                Bad Education   
...      ...      ...                                          ...   
8539   s8540  TV Show                                   The Tudors   
8557   s8558  TV Show                                The West Wing   
8684   s8685  TV Show                                      Vroomiz   
8712   s8713  TV Show                   Weird Wonders of the World   
8755   s8756  TV Show                            Women Behind Bars   

                     director  \
6066                      NaN   
6079  Jung-ah Im, Seung-uk Jo   
6174                      NaN   
6177                      NaN   
6213                      NaN   
...                       ...   
8539                      NaN   
8557                      NaN   
8684                      NaN   
8712                      NaN   
8755                      NaN   

                                                   cast  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...   
6079           Hyun-moo Jun, Si-kyung Sung, Se-yoon Yoo   
6174                                   Anthony Bourdain   
6177                                                NaN   
6213  Jack Whitehall, Mathew Horne, Sarah Solemani, ...   
...                                                 ...   
8539  Jonathan Rhys Meyers, Henry Cavill, James Frai...   
8557  Martin Sheen, Rob Lowe, Allison Janney, John S...   
8684  Joon-seok Song, Jeong-hwa Yang, Sang-hyun Um, ...   
8712                                      Chris Packham   
8755                                                NaN   

                                             country date_added  release_year  \
6066                                  United Kingdom        NaT          2013   
6079                                     South Korea        NaT          2017   
6174                                   United States        NaT          2018   
6177                                           Japan        NaT          2012   
6213                                  United Kingdom        NaT          2014   
...                                              ...        ...           ...   
8539  Ireland, Canada, United States, United Kingdom        NaT          2010   
8557                                   United States        NaT          2005   
8684                                     South Korea        NaT          2016   
8712                                  United Kingdom        NaT          2016   
8755                                   United States        NaT          2010   

     rating   duration                                          listed_in  \
6066  TV-MA  2 Seasons           British TV Shows, TV Comedies, TV Dramas   
6079  TV-PG  2 Seasons  International TV Shows, Korean TV Shows, Stand...   
6174  TV-PG  5 Seasons                                         Docuseries   
6177  TV-Y7  2 Seasons                             Anime Series, Kids' TV   
6213  TV-MA  3 Seasons                      British TV Shows, TV Comedies   
...     ...        ...                                                ...   
8539  TV-MA  4 Seasons                                          TV Dramas   
8557  TV-14  7 Seasons                                          TV Dramas   
8684   TV-Y  3 Seasons                          Kids' TV, Korean TV Shows   
8712  TV-PG  2 Seasons  British TV Shows, Docuseries, Science & Nature TV   
8755  TV-14  3 Seasons                         Crime TV Shows, Docuseries   

                                            description  year_added  \
6066  Set during the Russian Revolution, this comic ...         NaN   
6079  Led by a tri

Para que todas las entradas en `date_added` sean fechas no nulas y, consiguientemente, las columnas nuevas tengan valores no nulos, se puede introducir adrede una fecha ficticia en esta primera columna (y luego asignar las 2 últimas columnas como más arriba).

In [32]:
df['date_added'].sort_values()

5957   2008-01-01
6611   2008-02-04
5956   2009-05-05
5955   2009-11-18
7370   2010-11-01
          ...    
8539          NaT
8557          NaT
8684          NaT
8712          NaT
8755          NaT
Name: date_added, Length: 8807, dtype: datetime64[ns]

Por ejemplo, mi cumpleaños.

In [22]:
'2002-06-12 00:00:00' in df['date_added'].unique()

False

## Parte 2: Técnicas avanzadas de pandas

4. Utilizar `.loc` para seleccionar películas (`type == 'Movie'`) que fueron agregadas después del año 2018.

5. Utilizar `str.contains()` y `str.extract()`:

   * Filtrar títulos que contienen la palabra 'love' (sin distinguir mayúsculas/minúsculas).
   * Extraer la duración en minutos para las películas desde la columna `duration`.

6. Aplicar `explode()` sobre la columna `listed_in` para obtener una fila por cada género.

7. Obtener un top 10 de géneros más frecuentes utilizando `value_counts()`.

8. Aplicar `where()` y `mask()` para marcar las películas de más de 120 minutos como contenido largo en una nueva columna.

9. Utilizar `.loc` para filtrar películas que cumplen con:

   * Más de 100 minutos de duración.
   * Rating igual a `'R'`.
   * País igual a `'United States'`.

10. Utilizar `.style` para formatear visualmente el top 10 de películas más largas.

**Respuesta PARTE 2**

4. Utilizar `.loc` para seleccionar películas (`type == 'Movie'`) que fueron agregadas después del año 2018.

In [39]:
df.loc[(df['type'] == 'Movie') & (df['year_added'] > 2018)]#.head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country date_added  \
0                                       United States 2021-09-25   
6                                                 NaN 2021-09-24   
7   United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                       United States 2021-09-24   
12                            Germany, Czech Republic 2021-09-23   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  year_added  month_added  
0   As her father nears the end of his life, filmm...      2021.0          9.0  
6   Equestria's divided. But a bright-eyed hero be...      2021.0          9.0  
7   On a photo shoot in Ghana, an American model s...      2021.0          9.0  
9   A woman adjusting to life after a loss contend...      2021.0          9.0  
12  After most of her family is murdered in a terr...      2021.0          9.0

Notemos que la anterior lista no considera entradas cuyo `year_added` sea nulo.

In [51]:
df.loc[(df['type'] == 'Movie') & (df['year_added'] > 2018)]['year_added'].isnull().sum()

np.int64(0)

5. Utilizar `str.contains()` y `str.extract()`:

   * Filtrar títulos que contienen la palabra 'love' (sin distinguir mayúsculas/minúsculas).

In [55]:
df.loc[df['title'].str.contains('love', case=False, na=False)]#.head()

show_id     type                           title  \
25       s26  TV Show            Love on the Spectrum   
158     s159    Movie         Love Don't Cost a Thing   
159     s160    Movie                  Love in a Puff   
206     s207    Movie       LSD: Love, Sex Aur Dhokha   
227     s228    Movie                     Really Love   
...      ...      ...                             ...   
8567   s8568  TV Show  The Year of Happiness and Love   
8575   s8576  TV Show                 This Is My Love   
8649   s8650    Movie           Two Lovers and a Bear   
8721   s8722  TV Show                   What Is Love?   
8725   s8726    Movie            What's Up With Love?   

                       director  \
25                          NaN   
158                   Troy Byer   
159              Pang Ho-cheung   
206            Dibakar Banerjee   
227       Angel Kristi Williams   
...                         ...   
8567                        NaN   
8575                        NaN   
8649                 Kim Nguyen   
8721                        NaN   
8725  Rudy Soedjarwo, Riri Riza   

                                                   cast        country  \
25                                     Brooke Satchwell      Australia   
158   Nick Cannon, Christina Milian, Kenan Thompson,...  United States   
159   Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...      Hong Kong   
206   Nushrat Bharucha, Anshuman Jha, Neha Chauhan, ...          India   
227   Kofi Siriboe, Yootha Wong-Loi-Sing, Michael Ea...  United States   
...                                                 ...            ...   
8567  Amber Kuo, James Wen, Weber Yang, Sonia Sui, N...         Taiwan   
8575  Jin-mo Joo, Sa-rang Kim, Junior, Ja-in Lee, Su...    South Korea   
8649  Dane DeHaan, Tatiana Maslany, Gordon Pinsent, ...         Canada   
8721  Chris Wu, Jade Chou, Duncan Lai, King Chin, Ki...         Taiwan   
8725  Dian Sastrowardoyo, Nicholas Saputra, Ladya Ch...      Indonesia   

     date_added  release_year rating   duration  \
25   2021-09-21          2021  TV-14  2 Seasons   
158  2021-09-01          2003  PG-13    101 min   
159  2021-09-01          2010  TV-MA    103 min   
206  2021-08-27          2010  TV-MA    112 min   
227  2021-08-25          2020  TV-MA     95 min   
...         ...           ...    ...        ...   
8567 2016-09-01          2009  TV-PG   1 Season   
8575 2017-05-22          2015  TV-14   1 Season   
8649 2017-05-02          2016      R     95 min   
8721 2017-01-01          2012  TV-14   1 Season   
8725 2018-12-08          2002      R    110 min   

                                              listed_in  \
25       Docuseries, International TV Shows, Reality TV   
158                           Comedies, Romantic Movies   
159              Comedies, Dramas, International Movies   
206    Dramas, Independent Movies, International Movies   
227         Dramas, Independent Movies, Romantic Movies   
...                                                 ...   
8567  International TV Shows, Romantic TV Shows, TV ...   
8575  International TV Shows, Korean TV Shows, Roman...   
8649        Dramas, Independent Movies, Romantic Movies   
8721  International TV Shows, Romantic TV Shows, TV ...   
8725      Dramas, International Movies, Romantic Movies   

                                            description  year_added  \
25    Finding love can be hard for anyone. For young...      2021.0   
158   A nerdy teen tries to make himself cool by ass...      2021.0   
159   When the Hong Kong government enacts a ban on ...      2021.0   
206   This provocative drama examines how the voyeur...      2021.0   
227   A rising Black painter tries to break into a c...      2021.0   
...                                                 ...         ...   
8567  Jiang Chen-bo is a loser without prospects for...      2016.0   
8575  A renowned actor who is still pining over the ...      2017.0   
8649  A troubled couple living near the North Pol

5. Utilizar `str.contains()` y `str.extract()`:

   * Extraer la duración en minutos para las películas desde la columna `duration`.

Notemos que existen 3 películas sin duración registrada.

In [56]:
df.loc[df['type'] == 'Movie']['duration'].isnull().sum()

np.int64(3)

Hagamos una copia de `df` que incluya sólo a las películas y cambiemos la columna `duration` tal y como está por la duración en minutos.

In [102]:
filmes = df.loc[df['type'] == 'Movie']
filmes['duration'] = filmes['duration'].str.extract('(\d+)').astype(float)
filmes

<>:2: SyntaxWarning: invalid escape sequence '\d'
<>:2: SyntaxWarning: invalid escape sequence '\d'
/tmp/ipython-input-1197043478.py:2: SyntaxWarning: invalid escape sequence '\d'
  filmes['duration'] = filmes['duration'].str.extract('(\d+)').astype(float)
/tmp/ipython-input-1197043478.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filmes['duration'] = filmes['duration'].str.extract('(\d+)').astype(float)


show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country date_added  \
0                                         United States 2021-09-25   
6                                                   NaN 2021-09-24   
7     United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                         United States 2021-09-24   
12                              Germany, Czech Republic 2021-09-23   
...                                                 ...        ...   
8801                       United Arab Emirates, Jordan 2016-03-09   
8802                                      United States 2019-11-20   
8804                                      United States 2019-11-01   
8805                                      United States 2020-01-11   
8806                                              India 2019-03-02   

      release_year rating  duration  \
0             2020  PG-13      90.0   
6             2021     PG      91.0   
7             1993  TV-MA     125.0   
9             2021  PG-13     104.0   
12            2021  TV-MA     127.0   
...            ...    ...       ...   
8801          2015  TV-MA      96.0   
8802          2007      R     158.0   
8804          2009      R      88.0   
8805          2006     PG      88.0   
8806          2015  TV-14     111.0   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...      202

Notemos que este tratamiento no afecta a las entradas cuya duración es una entrada nula en el dataframe original (todas, coincidentemente, dirigidas/actuadas por Louis C.K.).

In [103]:
filmes.loc[filmes['duration'].isnull()]

show_id   type                                 title    director  \
5541   s5542  Movie                       Louis C.K. 2017  Louis C.K.   
5794   s5795  Movie                 Louis C.K.: Hilarious  Louis C.K.   
5813   s5814  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.   

            cast        country date_added  release_year  rating  duration  \
5541  Louis C.K.  United States 2017-04-04          2017  74 min       NaN   
5794  Louis C.K.  United States 2016-09-16          2010  84 min       NaN   
5813  Louis C.K.  United States 2016-08-15          2015  66 min       NaN   

     listed_in                                        description  year_added  \
5541    Movies  Louis C.K. muses on religion, eternal love, gi...      2017.0   
5794    Movies  Emmy-winning comedy writer Louis C.K. brings h...      2016.0   
5813    Movies  The comic puts his trademark hilarious/thought...      2016.0   

      month_added  
5541          4.0  
5794          9.0  
5813          8.0

6. Aplicar `explode()` sobre la columna `listed_in` para obtener una fila por cada género.

Hacemos una copia, separamos los distintos géneros `lirsted_in` considerando separador `', '`, y aplicamos el método `.explode()` para deglosar una misma entrada del df original por cada uno de sus géneros.

In [104]:
df_deglosado = df.copy()
df_deglosado['listed_in'] = df_deglosado['listed_in'].str.split(', ')
df_deglosado = df_deglosado.explode('listed_in')
df_deglosado#.head()

show_id     type                 title         director  \
0         s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show         Blood & Water              NaN   
1         s2  TV Show         Blood & Water              NaN   
1         s2  TV Show         Blood & Water              NaN   
2         s3  TV Show             Ganglands  Julien Leclercq   
...      ...      ...                   ...              ...   
8805   s8806    Movie                  Zoom     Peter Hewitt   
8805   s8806    Movie                  Zoom     Peter Hewitt   
8806   s8807    Movie                Zubaan      Mozez Singh   
8806   s8807    Movie                Zubaan      Mozez Singh   
8806   s8807    Movie                Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
...                                                 ...            ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

     date_added  release_year rating   duration                 listed_in  \
0    2021-09-25          2020  PG-13     90 min             Documentaries   
1    2021-09-24          2021  TV-MA  2 Seasons    International TV Shows   
1    2021-09-24          2021  TV-MA  2 Seasons                 TV Dramas   
1    2021-09-24          2021  TV-MA  2 Seasons              TV Mysteries   
2    2021-09-24          2021  TV-MA   1 Season            Crime TV Shows   
...         ...           ...    ...        ...                       ...   
8805 2020-01-11          2006     PG     88 min  Children & Family Movies   
8805 2020-01-11          2006     PG     88 min                  Comedies   
8806 2019-03-02          2015  TV-14    111 min                    Dramas   
8806 2019-03-02          2015  TV-14    111 min      International Movies   
8806 2019-03-02          2015  TV-14    111 min          Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...      2021.0   
1     After crossing paths at a party, a Cape Town t...      2021.0   
1     After crossing paths at a party, a Cape Town t...      2021.0   
1     After crossing paths at a party, a Cape Town t...      2021.0   
2     To protect his family from a powerful drug lor...      2021.0   
...                                                 ...         ...   
8805  Dragged from civilian life, a former superhero...      2020.0   
8805  Dragged from civilian life, a former superhero...      2020.0   
8806  A scrappy but poor boy worms his way into a ty...      2019.0   
8806  A scrappy but poor boy worms his way into a ty...      2019.0   
8806  A scrappy but poor boy worms his way into a ty...      2019.0   

      month_added  
0             9.0  
1             9.0  
1             9.0  
1             9.0  
2             9.0  
...           ...  
8805          1.0  
8805          1.0  
8806          3.0  
8806          3.0  
8806          3.0  

[19323 rows x 14 columns]

7. Obtener un top 10 de géneros más frecuentes utilizando `value_counts()`.

In [105]:
df_deglosado['listed_in'].value_counts().sort_values(ascending=False).head(10)

listed_in
International Movies        2752
Dramas                      2427
Comedies                    1674
International TV Shows      1351
Documentaries                869
Action & Adventure           859
TV Dramas                    763
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
Name: count, dtype: int64

8. Aplicar `where()` y `mask()` para marcar las películas de más de 120 minutos como contenido largo en una nueva columna.

Se ocupa el dataframe `filmes`, que tiene las películas (y sólo las películas) del dataframe original.

In [110]:
filmes['has_long_format'] = 0
filmes['has_long_format'] = filmes['has_long_format'].where(filmes['duration'] <= 120, True)
filmes['has_long_format'] = filmes['has_long_format'].mask(filmes['duration']  <= 120, False)
filmes#.head()

/tmp/ipython-input-1434738059.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filmes['has_long_format'] = 0
/tmp/ipython-input-1434738059.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filmes['has_long_format'] = filmes['has_long_format'].where(filmes['duration'] <= 120, True)
/tmp/ipython-input-1434738059.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: ht

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country date_added  \
0                                         United States 2021-09-25   
6                                                   NaN 2021-09-24   
7     United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                         United States 2021-09-24   
12                              Germany, Czech Republic 2021-09-23   
...                                                 ...        ...   
8801                       United Arab Emirates, Jordan 2016-03-09   
8802                                      United States 2019-11-20   
8804                                      United States 2019-11-01   
8805                                      United States 2020-01-11   
8806                                              India 2019-03-02   

      release_year rating  duration  \
0             2020  PG-13      90.0   
6             2021     PG      91.0   
7             1993  TV-MA     125.0   
9             2021  PG-13     104.0   
12            2021  TV-MA     127.0   
...            ...    ...       ...   
8801          2015  TV-MA      96.0   
8802          2007      R     158.0   
8804          2009      R      88.0   
8805          2006     PG      88.0   
8806          2015  TV-14     111.0   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...      202

9. Utilizar `.loc` para filtrar películas que cumplen con:

   * Más de 100 minutos de duración.
   * Rating igual a `'R'`.
   * País igual a `'United States'`.

Se ocupa dataframe `filmes`, que tiene las películas (y sólo las películas), y cuya columna `duración` es un float (salvo cuando es nulo, donde es `NaN`). No se consideran casos donde `'United States'` pudiese estar mal escrito.

In [111]:
filmes.loc[(filmes['duration']>100) & (filmes['rating']=='R') & (filmes['country']=='United States')]#.head()

show_id   type                        title               director  \
48       s49  Movie                 Training Day          Antoine Fuqua   
81       s82  Movie                         Kate  Cedric Nicolas-Troyan   
131     s132  Movie  Blade Runner: The Final Cut           Ridley Scott   
139     s140  Movie           Do the Right Thing              Spike Lee   
144     s145  Movie                  House Party        Reginald Hudlin   
...      ...    ...                          ...                    ...   
8678   s8679  Movie              Vincent N Roxxy   Gary Michael Schultz   
8691   s8692  Movie                    Wakefield          Robin Swicord   
8751   s8752  Movie              Wish I Was Here             Zach Braff   
8754   s8755  Movie                       Wolves        Bart Freundlich   
8802   s8803  Movie                       Zodiac          David Fincher   

                                                   cast        country  \
48    Denzel Washington, Ethan Hawke, Scott Glenn, T...  United States   
81    Mary Elizabeth Winstead, Jun Kunimura, Woody H...  United States   
131   Harrison Ford, Rutger Hauer, Sean Young, Edwar...  United States   
139   Danny Aiello, Ossie Davis, Ruby Dee, Richard E...  United States   
144   Christopher Reid, Christopher Martin, Robin Ha...  United States   
...                                                 ...            ...   
8678  Emile Hirsch, Zoë Kravitz, Emory Cohen, Zoey D...  United States   
8691  Bryan Cranston, Jennifer Garner, Jason O'Mara,...  United States   
8751  Zach Braff, Kate Hudson, Donald Faison, Joey K...  United States   
8754  Michael Shannon, Carla Gugino, Taylor John Smi...  United States   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   

     date_added  release_year rating  duration  \
48   2021-09-16          2001      R     122.0   
81   2021-09-10          2021      R     106.0   
131  2021-09-01          1982      R     117.0   
139  2021-09-01          1989      R     120.0   
144  2021-09-01          1990      R     104.0   
...         ...           ...    ...       ...   
8678 2017-09-02          2016      R     101.0   
8691 2019-03-02          2016      R     109.0   
8751 2018-08-16          2014      R     106.0   
8754 2019-03-29          2016      R     109.0   
8802 2019-11-20          2007      R     158.0   

                                            listed_in  \
48                                  Dramas, Thrillers   
81                                 Action & Adventure   
131   Action & Adventure, Classic Movies, Cult Movies   
139                  Classic Movies, Comedies, Dramas   
144                             Comedies, Cult Movies   
...                                               ...   
8678                                Dramas, Thrillers   
8691                                           Dramas   
8751             Comedies, Dramas, Independent Movies   
8754        Dramas, Independent Movies, Sports Movies   
8802                   Cult Movies, Dramas, Thrillers   

                                            description  year_added  \
48    A rookie cop with one day to prove himself to ...      2021.0   
81    Slipped a fatal poison on her final job, a rut...      2021.0   
131   In a smog-choked dystopian Los Angeles, blade ...      2021.0   
139   On a sweltering day in Brooklyn, simmering rac...      2021.0   
144   Grounded by his strict father, Kid risks life ...      2021.0   
...                                                 ...         ...   
8678  In rural Louisiana, a terse loner forges a red...      2017.0   
8691  An unhappy father and lawyer quits his suburba...      2019.0   
8751  With his acting career moribund, Aidan Bloom s...      2018.0   
8754  A promising high school basketball player has ...      2019.0   
8802  A political cartoonist, a crime reporter and a...      2019.0   

      month_added has_long_format  
48            9.0            True  
81  

10. Utilizar `.style` para formatear visualmente el top 10 de películas más largas.

In [119]:
by_longest = filmes.copy()
by_longest = by_longest.sort_values(by='duration', ascending=False).head(10)
by_longest[['title', 'director', 'cast', 'duration', 'rating', 'country']].style.background_gradient(subset=['duration'], cmap='Reds')



### Pregunta Desafío

11. ¿Cuáles son las combinaciones más frecuentes de género y rating en el dataset?
    (Sugerencia: utilizar `value_counts` con `subset=["genre", "rating"]` después de aplicar `explode()`).



### Bonus: Análisis de duplicados y limpieza

12. ¿Existen películas con el mismo nombre (`title`) pero con distinto año de lanzamiento (`release_year`)?
13. ¿Cuántos títulos únicos hay en total en la columna `title`?





**Respuesta DESAFÍO & BONUS**

11. ¿Cuáles son las combinaciones más frecuentes de género y rating en el dataset?
    (Sugerencia: utilizar `value_counts` con `subset=["genre", "rating"]` después de aplicar `explode()`).

In [121]:
df_deglosado.value_counts(subset=['listed_in','rating'])#.head()

listed_in               rating  
International Movies    TV-MA       1130
                        TV-14       1065
Dramas                  TV-MA        830
International TV Shows  TV-MA        714
Dramas                  TV-14        693
                                    ... 
TV Sci-Fi & Fantasy     NR             1
TV Mysteries            TV-G           1
TV Sci-Fi & Fantasy     TV-Y7-FV       1
TV Shows                R              1
Action & Adventure      G              1
Name: count, Length: 309, dtype: int64

Las 3 combinaciones más populares son:
1. International Movies, TV-MA
2. International Movies, TV-14
3. Dramas, TV-MA

12. ¿Existen películas con el mismo nombre (`title`) pero con distinto año de lanzamiento (`release_year`)?

No, porque, como se ve abajo, existen 8807 títulos distintos; esto es, precisamente la cantidad de entradas, de modo que no hay dos películas con igual título para empezar.


13. ¿Cuántos títulos únicos hay en total en la columna `title`?

In [120]:
df['title'].nunique()

8807